Netty 4 server fixes

lukiano opened the pull request:
- Fix casting bug at shutdown. 

- Call retain() on ByteBuf before offering to an AsyncQueue

vkostyukov commented on finagle-netty4/src/main/scala/com/twitter/finagle/netty4/Netty4Listener.scala:
```diff
@@ -128,11 +128,16 @@ private[finagle] case class Netty4Listener[In, Out](
-          .addListener(new GenericFutureListener[Nothing] {
-            def operationComplete(future: Nothing): Unit = p.setDone()
-          })
+        def shutdownListener[N <: NettyFuture[_]]: GenericFutureListener[N] =
```
ffti: `NettyFuture[Any]` and `GenericFutureListener[Any]` should work as well.

I have an internal patch that also fixes that but it's been soaking for review for a long time so I'm totally fine to merge this in.

vkostyukov commented on finagle-netty4/src/main/scala/com/twitter/finagle/netty4/channel/Netty4ServerChannelInitializer.scala:
```diff
@@ -103,13 +104,25 @@ private[netty4] class Netty4ServerChannelInitializer(
   }
 }
 
+private[netty4] object ContentRetainer extends ChannelInboundHandlerAdapter {
```
Ah that's interesting! I guess you're right and we should totally ratain those objects.

What do you think about fixing `ChannelTransport` to do `ReferenceCountUtil.ratain(msg)` before offering the message into `AsyncQueue`, instead of introducing a new handler here?

lukiano replied:
I thought of that as well, but wasn't sure about maybe mixing concerns

dschobel commented:
thanks @lukiano, can you add a unit test for the casting bug and updating the git commit to follow https://github.com/twitter/finagle/blob/develop/CONTRIBUTING.md ?

Taking a step back from the question of the reference leak this patches for pipelines that let undecoded buffers through, I don't think letting netty bytebufs leak into the finagle layer of transports and dispatchers is something we should encourage so I'd rather install a terminal bytebuf 2 `com.twitter.util.Buf` handler and then ask protocol implementors who want to use bytebufs and the netty convenience classes to build their protocols in a channel handler. 

That lets us sequester the netty types to the pipeline and nudges people towards using buf and gaining the benefit of not being tied to a specific version of netty.

lukiano commented:
@dschobel Will do!
About your second paragraph, this isn't just ByteBuf, but actual Netty messages (HttpContent and inheritance) which also extends ByteBufHolder, which itself extends ReferenceCounted, so we can call retain()
Related, eventually this will need to be improved to support pooled ByteBufs

mosesn commented:
@lukiano yah, we were thinking about how to implement pooled ByteBufs, we were thinking we could use PhantomReferences?  What do you think?

lukiano commented:
@mosesn Not sure yet as I'm not entirely familiar with ByteBufs, but was thinking something along the lines of when the Buf readerIndex reaches the end, then release the object.

mosesn commented:
@lukiano but what if we want to read it again, or we don't want to finish reading it?

dschobel commented:
> About your second paragraph, this isn't just ByteBuf, but actual Netty messages (HttpContent and inheritance) which also extends ByteBufHolder, which itself extends ReferenceCounted, so we can call retain()
> Related, eventually this will need to be improved to support pooled ByteBufs

@lukiano thinking about it a bit more, since this never releases buffers it's not usable with the pooled allocator unless the protocol implementation copies every buffer, in which case, why use the pooled allocator?

It seems like it would be safer and less confusing to let netty's leak detectors blow up early if someone tries to use the pooled allocator rather than force that constraint on protocol implementors.

As @mosesn says,  there are options for making pooled work without copying but that will be a bigger effort.

lukiano commented:
I think the user should be able to choose between non-pooled or pooled buffers. Also they should be able to choose if they are heap or direct based. I think Netty defaults to direct, but in case of Unpooled buffers, maybe heap ones are better as a default.
If the user chooses to go in the direct/pooled route, it must be aware of the compromises. These are examples:

- If there's only one buffer, like a short message, or a chunked message put together via HttpObjectAggregator, then it will be released when the service responds. That is, Future[Response] is completed.
- If there are chunk of buffers, provided by the `com.twitter.io.Reader` of the request, each time `read(n: Int)` is called to get a new chunk, the previous one is released.

I'm sure there are corner cases. This PR is to make it work with unpooled buffers

dschobel commented:
@lukiano okay I was missing that unpooled buffers can be freed and disabling the leak detection code doesn't stop it from being a runtime exception to access its contents.

I wrote two tests for the listener and the transporter with no protocol implementation and the listener's buffers hit finagle space with `refCnt == 1` which seems correct.

There is a bug on the transporter side because the `ChannelTransport`'s SimpleChannelHandler instance auto-releases its buffers and will cause a runtime failure if the protocol implementation doesn't explicitly retain them.

lukiano commented:
@dschobel I have updated this PR with the SimpleChannelHandler change, updated the commit messages, and updated the tests to reflect the casting issue.

mosesn commented:
LGTM

dschobel commented:
thanks @lukiano, I'm merging it internally. it will sync out to the develop branch on Monday/Tuesday.

dschobel commented:
merged in 2f1786b4658c60a0f22dafc83b0cdfcccba5619d. thanks @lukiano
